### PR TITLE
Modify Skip GAN connection

### PR DESCRIPTION
**Pull request by Manasi1999**
- Add sigmoid to skip GAN weight
- Create new training function removing the forced constraint

### CODE REVIEW

File: src/network.py [hunk 1]
```diff
@@ -171,4 +171,75 @@ def forward(self, x):
...
+        
+    def forward(self, input):
+        x, y = input.shape
+        if y != self.in_features:
+            print(f'Wrong Input Features. Please use tensor with {self.in_features} Input Features')
+            return 0
+        output = torch.Tensor(x,1)
+        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        for i in range(x):
+          w = torch.nn.sigmoid(self.weight)
```
Comment (dhavala) [inline src/network.py]: w = torch.nn.sigmoid(self.weight)
this can be defined outside the for loop, since this is constant inside the for loop 

but o.w looks good

File: src/network.py [hunk 2]
```diff
@@ -171,4 +171,75 @@ def forward(self, x):
...
+        self.weight = torch.nn.Parameter(torch.Tensor(1,1))
+        if bias:
+            self.bias = torch.nn.Parameter(torch.Tensor(out_features))
+        else:
+            self.register_parameter('bias', None)
+        self.reset_parameters()
+
+    #Initialise the weights  
+    def reset_parameters(self):
+        torch.nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))
```
Comment (dhavala) [inline src/network.py]: initiate the self.weight to 0 and pass it as a parameter. 
when creating skip node pass init=0

do not use random initialization

File: src/network.py [hunk 3]
```diff
@@ -171,4 +171,75 @@ def forward(self, x):
...
+    x = self.relu(x)
+    x = self.hidden2(x)
+    x = self.relu(x)
+    x = self.hidden2(x)
+    x = self.relu(x)
+    x = self.hidden2(x)
+    x = self.relu(x)
+    y_gan = self.output(x)
+    out = torch.cat((y_gan , y_abc),1)
+    out = self.skipNode(out)
```
Comment (dhavala) [inline src/network.py:245]: yes, this looks correct. since now ouput is correctly (weighed average), we do not need to apply  clamp in the training routine

File: src/ABC_train_test4.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      inputs_fake = torch.cat((x_batch_cuda,generated_y),dim=1).to(device)
+      disc_fake_pred = disc(inputs_fake)
+
+      gen_loss = criterion(disc_fake_pred,real_labels)
+      epoch_loss_gen.append(gen_loss.item())
+
+      #Update gradients 
+      gen_loss.backward()
+      #Update optimizer 
+      gen_opt.step()
```
Comment (dhavala) [inline src/ABC_train_test4.py:157]: yes, no need to apply clamp the skip weight. it is handled in the model itself.
looks good.

**Review by dhavala — COMMENTED** [review]
intialize the skip weight in the model to zero, instead of random!

File: src/network.py (hunk 1 above)
Comment (Manasi1999) [reply src/network.py]: Done

**Comment by Manasi1999** [comment]
> intialize the skip weight in the model to zero, instead of random!

Done
